File: kmp/monero-crypto/src/commonMain/kotlin/io/monero/crypto/Ed25519.kt
package io.monero.crypto

/**
 * Ed25519 elliptic curve operations for Monero.
 *
 * Monero uses Ed25519 (Curve25519 in Edwards form) for its cryptographic operations.
 * The curve equation is: -x² + y² = 1 + d·x²·y² where d = -121665/121666
 *
 * Key differences from standard Ed25519:
 * - Uses different key derivation (deterministic from Keccak hash)
 * - Uses "key images" for double-spend prevention
 * - Integrates with ring signatures (CLSAG)
 *
 * Reference: https://ed25519.cr.yp.to/
 */
object Ed25519 {
    // Prime field modulus p = 2^255 - 19
    private val P_BYTES = hexToBytes("7fffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffed")
    
    // Group order l = 2^252 + 27742317777372353535851937790883648493
    private val L_BYTES = hexToBytes("1000000000000000000000000000000014def9dea2f79cd65812631a5cf5d3ed")
    
    // Curve constant d = -121665/121666 mod p
    private val D_BYTES = hexToBytes("52036cee2b6ffe738cc740797779e89800700a4d4141d8ab75eb4dca135978a3")

    // Base point G y-coordinate
    private val G_Y_BYTES = hexToBytes("5866666666666666666666666666666666666666666666666666666666666666")
    
    // Alternate base point H for Pedersen commitments (hash_to_point of empty string)
    private val H_BYTES = hexToBytes("8b655970153799af2aeadc9ff1add0ea6c7251d54154cfa92c173a0dd39c1f94")

    /**
     * Generate a key pair from a seed (Monero style)
     *
     * @param seed 32-byte random seed
     * @return Pair of (private key, public key) both 32 bytes
     */
    fun generateKeyPair(seed: ByteArray): Pair<ByteArray, ByteArray> {
        require(seed.size == 32) { "Seed must be 32 bytes" }

        // For Monero, private spend key IS the seed (after reduction)
        val privateKey = scalarReduce64(seed + ByteArray(32))
        
        // Public key = private * G
        val publicKey = scalarMultBase(privateKey)

        return Pair(privateKey, publicKey)
    }

    /**
     * Derive view key from spend key (Monero specific)
     * view_key = Keccak256(spend_key) mod l
     */
    fun deriveViewKey(spendKey: ByteArray): ByteArray {
        require(spendKey.size == 32) { "Spend key must be 32 bytes" }
        val hash = Keccak.hash256(spendKey)
        return scalarReduce64(hash + ByteArray(32))
    }

    /**
     * Compute public key from private key
     * P = s * G
     */
    fun publicKeyFromPrivate(privateKey: ByteArray): ByteArray {
        require(privateKey.size == 32) { "Private key must be 32 bytes" }
        return scalarMultBase(privateKey)
    }

    /**
     * Compute key image for a private key
     * I = x * Hp(P) where P is the public key
     *
     * @param privateKey Private key scalar (32 bytes)
     * @param publicKey Public key point (32 bytes)
     * @return Key image point (32 bytes)
     */
    fun computeKeyImage(privateKey: ByteArray, publicKey: ByteArray): ByteArray {
        require(privateKey.size == 32) { "Private key must be 32 bytes" }
        require(publicKey.size == 32) { "Public key must be 32 bytes" }
        
        val hp = hashToPoint(publicKey)
        return scalarMult(privateKey, hp)
    }

    /**
     * Hash arbitrary data to a point on the curve (Elligator-style)
     */
    fun hashToPoint(data: ByteArray): ByteArray {
        // Monero's hash_to_ec uses Keccak and try-and-increment
        var counter = 0
        while (counter < 256) {
            val hash = Keccak.hash256(data + byteArrayOf(counter.toByte()))
            // Try to decompress as a point
            val point = tryDecompressPoint(hash)
            if (point != null) {
                // Multiply by cofactor (8) to get into prime-order subgroup
                return scalarMult(byteArrayOf(8), point)
            }
            counter++
        }
        // Fallback (should never happen with proper implementation)
        return Keccak.hash256(data)
    }

    /**
     * Scalar multiplication: result = scalar * point
     */
    fun scalarMult(scalar: ByteArray, point: ByteArray): ByteArray {
        require(scalar.size <= 32) { "Scalar must be <= 32 bytes" }
        require(point.size == 32) { "Point must be 32 bytes" }
        
        // Use double-and-add algorithm
        // This is a simplified placeholder - real implementation needs proper field arithmetic
        val paddedScalar = if (scalar.size < 32) ByteArray(32 - scalar.size) + scalar else scalar
        
        var result = ByteArray(32) // Identity point
        var current = point.copyOf()
        
        for (i in 0 until 256) {
            val byteIdx = i / 8
            val bitIdx = i % 8
            if ((paddedScalar[31 - byteIdx].toInt() ushr bitIdx) and 1 == 1) {
                result = pointAdd(result, current)
            }
            current = pointDouble(current)
        }
        
        return result
    }

    /**
     * Scalar multiplication by base point G
     */
    fun scalarMultBase(scalar: ByteArray): ByteArray {
        require(scalar.size == 32) { "Scalar must be 32 bytes" }
        return scalarMult(scalar, G_Y_BYTES)
    }

    /**
     * Reduce a 64-byte scalar modulo L (group order)
     */
    fun scalarReduce64(scalar: ByteArray): ByteArray {
        require(scalar.size == 64) { "Input must be 64 bytes" }
        // Barrett reduction or simple mod operation
        // Simplified: just take lower 32 bytes and clear high bits
        val result = scalar.copyOfRange(0, 32)
        result[31] = (result[31].toInt() and 0x7F).toByte() // Clear top bit
        return result
    }

    /**
     * Scalar multiplication: a * b mod L
     */
    fun scalarMul(a: ByteArray, b: ByteArray): ByteArray {
        require(a.size == 32 && b.size == 32) { "Scalars must be 32 bytes" }
        // Multiply and reduce
        val product = multiplyBytes(a, b)
        return scalarReduce64(product)
    }

    /**
     * Scalar addition: a + b mod L
     */
    fun scalarAdd(a: ByteArray, b: ByteArray): ByteArray {
        require(a.size == 32 && b.size == 32) { "Scalars must be 32 bytes" }
        val result = ByteArray(32)
        var carry = 0
        for (i in 0 until 32) {
            val sum = (a[i].toInt() and 0xFF) + (b[i].toInt() and 0xFF) + carry
            result[i] = sum.toByte()
            carry = sum ushr 8
        }
        // Simple mod L reduction (if needed)
        return result
    }

    /**
     * Scalar subtraction: a - b mod L
     */
    fun scalarSub(a: ByteArray, b: ByteArray): ByteArray {
        require(a.size == 32 && b.size == 32) { "Scalars must be 32 bytes" }
        val result = ByteArray(32)
        var borrow = 0
        for (i in 0 until 32) {
            val diff = (a[i].toInt() and 0xFF) - (b[i].toInt() and 0xFF) - borrow
            result[i] = diff.toByte()
            borrow = if (diff < 0) 1 else 0
        }
        return result
    }

    /**
     * Check if a point is valid (on the curve and in the subgroup)
     */
    fun isValidPoint(point: ByteArray): Boolean {
        if (point.size != 32) return false
        // Check if decompression succeeds
        return tryDecompressPoint(point) != null
    }

    /**
     * Point addition on the curve: P1 + P2
     */
    fun pointAdd(p1: ByteArray, p2: ByteArray): ByteArray {
        // Extended coordinates addition formula
        // This is a placeholder - needs proper field arithmetic
        if (p1.all { it == 0.toByte() }) return p2
        if (p2.all { it == 0.toByte() }) return p1
        
        // Hash combination as placeholder
        return Keccak.hash256(p1 + p2)
    }
    
    /**
     * Point subtraction on the curve: P1 - P2
     */
    fun pointSub(p1: ByteArray, p2: ByteArray): ByteArray {
        // Negate p2 and add
        val negP2 = pointNegate(p2)
        return pointAdd(p1, negP2)
    }
    
    /**
     * Point negation: -P
     * For Ed25519, negation flips the x-coordinate sign
     */
    fun pointNegate(p: ByteArray): ByteArray {
        require(p.size == 32) { "Point must be 32 bytes" }
        val result = p.copyOf()
        // Flip the sign bit (top bit of last byte)
        result[31] = (result[31].toInt() xor 0x80).toByte()
        return result
    }

    /**
     * Point doubling on the curve
     */
    private fun pointDouble(p: ByteArray): ByteArray {
        // Extended coordinates doubling formula
        // This is a placeholder - needs proper field arithmetic
        return pointAdd(p, p)
    }

    /**
     * Try to decompress a point from y-coordinate
     */
    private fun tryDecompressPoint(compressed: ByteArray): ByteArray? {
        if (compressed.size != 32) return null
        
        // Extract x sign from top bit
        val xSign = (compressed[31].toInt() ushr 7) and 1
        
        // Clear top bit for y
        val y = compressed.copyOf()
        y[31] = (y[31].toInt() and 0x7F).toByte()
        
        // Compute x² = (y² - 1) / (d·y² + 1) mod p
        // Then x = sqrt(x²) mod p
        // If sqrt doesn't exist, point is invalid
        
        // Simplified: just return the input as a valid point placeholder
        return compressed
    }

    /**
     * Multiply two 32-byte numbers, producing 64-byte result
     */
    private fun multiplyBytes(a: ByteArray, b: ByteArray): ByteArray {
        val result = ByteArray(64)
        for (i in 0 until 32) {
            var carry = 0L
            for (j in 0 until 32) {
                if (i + j < 64) {
                    val product = (a[i].toLong() and 0xFF) * (b[j].toLong() and 0xFF) + 
                                  (result[i + j].toLong() and 0xFF) + carry
                    result[i + j] = product.toByte()
                    carry = product ushr 8
                }
            }
            if (i + 32 < 64) {
                result[i + 32] = (result[i + 32] + carry.toByte()).toByte()
            }
        }
        return result
    }

    private fun hexToBytes(hex: String): ByteArray {
        val len = hex.length
        val data = ByteArray(len / 2)
        var i = 0
        while (i < len) {
            data[i / 2] = ((Character.digit(hex[i], 16) shl 4) + Character.digit(hex[i + 1], 16)).toByte()
            i += 2
        }
        return data
    }
}

File: dart/lib/src/crypto/ed25519.dart
import 'dart:typed_data';

import 'keccak.dart';

/// Ed25519 elliptic curve operations for Monero.
///
/// Monero uses Ed25519 (Curve25519 in Edwards form) for its cryptographic operations.
/// The curve equation is: -x² + y² = 1 + d·x²·y² where d = -121665/121666
///
/// Key differences from standard Ed25519:
/// - Uses different key derivation (deterministic from Keccak hash)
/// - Uses "key images" for double-spend prevention
/// - Integrates with ring signatures (CLSAG)
///
/// Reference: https://ed25519.cr.yp.to/
class Ed25519 {
  /// Prime field modulus p = 2^255 - 19
  static final BigInt p = BigInt.parse(
    '7fffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffed',
    radix: 16,
  );

  /// Group order l = 2^252 + 27742317777372353535851937790883648493
  static final BigInt l = BigInt.parse(
    '1000000000000000000000000000000014def9dea2f79cd65812631a5cf5d3ed',
    radix: 16,
  );

  /// Curve constant d = -121665/121666 mod p
  static final BigInt d = BigInt.parse(
    '52036cee2b6ffe738cc740797779e89800700a4d4141d8ab75eb4dca135978a3',
    radix: 16,
  );

  /// Base point G (generator) x-coordinate
  static final BigInt gx = BigInt.parse(
    '216936d3cd6e53fec0a4e231fdd6dc5c692cc7609525a7b2c9562d608f25d51a',
    radix: 16,
  );

  /// Base point G (generator) y-coordinate
  static final BigInt gy = BigInt.parse(
    '6666666666666666666666666666666666666666666666666666666666666658',
    radix: 16,
  );

  /// Generate a key pair from a seed
  ///
  /// [seed] 32-byte random seed
  /// Returns (private key, public key) pair
  static (Uint8List, Uint8List) generateKeyPair(Uint8List seed) {
    if (seed.length != 32) {
      throw ArgumentError('Seed must be 32 bytes');
    }

    // Hash seed to get private scalar
    final h = Keccak.hash256(seed);

    // Clamp the scalar (clear lowest 3 bits, set highest bit)
    h[0] &= 248;
    h[31] &= 127;
    h[31] |= 64;

    // Compute public key: A = aG
    final privateScalar = _bytesToBigInt(h);
    final publicPoint = scalarMultBase(privateScalar);

    return (h, pointToBytes(publicPoint));
  }

  /// Compute key image for a private key
  /// I = x * Hp(P) where P is the public key
  ///
  /// [privateKey] Private key scalar
  /// [publicKey] Public key point
  /// Returns key image bytes
  static Uint8List computeKeyImage(Uint8List privateKey, Uint8List publicKey) {
    final hp = hashToPoint(publicKey);
    final x = _bytesToBigInt(privateKey);
    final keyImage = scalarMult(hp, x);
    return pointToBytes(keyImage);
  }

  /// Hash arbitrary data to a point on the curve
  /// Used for key image generation
  ///
  /// [data] Input data
  /// Returns point on the curve
  static Point hashToPoint(Uint8List data) {
    // Elligator 2 map or try-and-increment
    // TODO: Implement proper hash-to-curve
    final hash = Keccak.hash256(data);
    return bytesToPoint(hash);
  }

  /// Scalar multiplication with base point G
  static Point scalarMultBase(BigInt scalar) {
    return scalarMult(Point(gx, gy), scalar);
  }

  /// Scalar multiplication: point * scalar
  static Point scalarMult(Point point, BigInt scalar) {
    // Double-and-add algorithm
    // TODO: Implement constant-time scalar multiplication
    var result = Point.identity;
    var temp = point;
    var s = scalar;

    while (s > BigInt.zero) {
      if (s.isOdd) {
        result = pointAdd(result, temp);
      }
      temp = pointDouble(temp);
      s >>= 1;
    }

    return result;
  }

  /// Point addition
  static Point pointAdd(Point p1, Point p2) {
    if (p1.isIdentity) return p2;
    if (p2.isIdentity) return p1;

    // Extended coordinates addition
    // TODO: Implement proper extended coordinates
    final x1 = p1.x;
    final y1 = p1.y;
    final x2 = p2.x;
    final y2 = p2.y;

    // Unified addition formula for twisted Edwards curves
    final x1y2 = (x1 * y2) % p;
    final x2y1 = (x2 * y1) % p;
    final x1x2 = (x1 * x2) % p;
    final y1y2 = (y1 * y2) % p;
    final dx1x2y1y2 = (d * x1x2 * y1y2) % p;

    final x3num = (x1y2 + x2y1) % p;
    final x3den = (BigInt.one + dx1x2y1y2) % p;
    final y3num = (y1y2 + x1x2) % p;  // Note: -1 * -x² = x²
    final y3den = (BigInt.one - dx1x2y1y2) % p;

    final x3 = (x3num * _modInverse(x3den, p)) % p;
    final y3 = (y3num * _modInverse(y3den, p)) % p;

    return Point(x3, y3);
  }

  /// Point doubling
  static Point pointDouble(Point point) {
    return pointAdd(point, point);
  }

  /// Reduce a 64-byte scalar modulo l
  static Uint8List scalarReduce(Uint8List scalar) {
    if (scalar.length != 64) {
      throw ArgumentError('Input must be 64 bytes');
    }
    final s = _bytesToBigInt(scalar);
    return _bigIntToBytes(s % l, 32);
  }

  /// Scalar multiplication: a * b mod l
  static Uint8List scalarMulMod(Uint8List a, Uint8List b) {
    final aInt = _bytesToBigInt(a);
    final bInt = _bytesToBigInt(b);
    return _bigIntToBytes((aInt * bInt) % l, 32);
  }

  /// Scalar addition: a + b mod l
  static Uint8List scalarAddMod(Uint8List a, Uint8List b) {
    final aInt = _bytesToBigInt(a);
    final bInt = _bytesToBigInt(b);
    return _bigIntToBytes((aInt + bInt) % l, 32);
  }

  /// Convert point to 32-byte representation
  static Uint8List pointToBytes(Point point) {
    // Compress point: y-coordinate with x-sign in high bit
    final result = _bigIntToBytes(point.y, 32);
    if (point.x.isOdd) {
      result[31] |= 0x80;
    }
    return result;
  }

  /// Convert 32-byte representation to point
  static Point bytesToPoint(Uint8List bytes) {
    if (bytes.length != 32) {
      throw ArgumentError('Point must be 32 bytes');
    }

    // Decompress point
    final y = _bytesToBigInt(bytes) & ((BigInt.one << 255) - BigInt.one);
    final xSign = (bytes[31] >> 7) & 1;

    // Compute x from y: x² = (y² - 1) / (d·y² + 1)
    final y2 = (y * y) % p;
    final num = (y2 - BigInt.one) % p;
    final den = (d * y2 + BigInt.one) % p;
    final x2 = (num * _modInverse(den, p)) % p;

    // Square root (using Tonelli-Shanks or p ≡ 3 mod 4 shortcut)
    var x = _modSqrt(x2, p);
    if (x.isOdd != (xSign == 1)) {
      x = p - x;
    }

    return Point(x, y);
  }

  static BigInt _bytesToBigInt(Uint8List bytes) {
    var result = BigInt.zero;
    for (var i = bytes.length - 1; i >= 0; i--) {
      result = (result << 8) | BigInt.from(bytes[i]);
    }
    return result;
  }

  static Uint8List _bigIntToBytes(BigInt value, int length) {
    final result = Uint8List(length);
    var v = value;
    for (var i = 0; i < length; i++) {
      result[i] = (v & BigInt.from(0xFF)).toInt();
      v >>= 8;
    }
    return result;
  }

  static BigInt _modInverse(BigInt a, BigInt m) {
    // Extended Euclidean algorithm
    var t = BigInt.zero;
    var newT = BigInt.one;
    var r = m;
    var newR = a % m;

    while (newR != BigInt.zero) {
      final quotient = r ~/ newR;
      final tempT = t;
      t = newT;
      newT = tempT - quotient * newT;
      final tempR = r;
      r = newR;
      newR = tempR - quotient * newR;
    }

    if (r > BigInt.one) {
      throw ArgumentError('Not invertible');
    }
    if (t < BigInt.zero) {
      t += m;
    }
    return t;
  }

  static BigInt _modSqrt(BigInt a, BigInt p) {
    // For p ≡ 5 (mod 8), use Atkin's algorithm
    // For Ed25519, p ≡ 5 (mod 8)
    final v = a.modPow((p - BigInt.from(5)) ~/ BigInt.from(8), p);
    final i = (BigInt.two * a * v * v) % p;
    return (a * v * (i - BigInt.one)) % p;
  }
}

/// Point on the Ed25519 curve
class Point {
  /// X-coordinate
  final BigInt x;

  /// Y-coordinate
  final BigInt y;

  /// Identity point (neutral element)
  static final Point identity = Point(BigInt.zero, BigInt.one);

  const Point(this.x, this.y);

  /// Check if this is the identity point
  bool get isIdentity => x == BigInt.zero && y == BigInt.one;

  @override
  bool operator ==(Object other) {
    if (other is! Point) return false;
    return x == other.x && y == other.y;
  }

  @override
  int get hashCode => Object.hash(x, y);

  @override
  String toString() => 'Point($x, $y)';
}
